File: react-frontend/src/services/ComplaintService.js
import api from "./api";

// Create Complaint
export const createComplaint = (complaint) => {
    return api.post("/api/complaints", complaint);
};

// Get All Complaints
export const getAllComplaints = () => {
    return api.get("/api/complaints");
};

// Get Complaint By ID
export const getComplaintById = (id) => {
    return api.get(`/api/complaints/${id}`);
};

// Get Complaints By User
export const getComplaintsByUser = (userId) => {
    return api.get(`/api/complaints/user/${userId}`);
};

// Search Complaints By Title
export const searchComplaint = (title) => {
    return api.get(`/api/complaints/search?title=${title}`);
};

// Filter Complaints By Status
export const filterComplaint = (status) => {
    return api.get(`/api/complaints/status?status=${status}`);
};

// Update Complaint Status
export const updateComplaintStatus = (id, status) => {
    return api.put(`/api/complaints/${id}?status=${status}`);
};

// Delete Complaint
export const deleteComplaint = (id) => {
    return api.delete(`/api/complaints/${id}`);
};

File: backend/src/main/java/com/smartcomplaint/backend/service/ComplaintService.java
package com.smartcomplaint.backend.service;

import com.smartcomplaint.backend.entity.Complaint;
import com.smartcomplaint.backend.repository.ComplaintRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ComplaintService {

    @Autowired
    private ComplaintRepository complaintRepository;

    // Save Complaint
    public Complaint saveComplaint(Complaint complaint) {
        complaint.setStatus("Pending");
        return complaintRepository.save(complaint);
    }

    // Get All Complaints
    public List<Complaint> getAllComplaints() {
        return complaintRepository.findAll();
    }

    // Get Complaint By Id
    public Complaint getComplaintById(Long id) {
        return complaintRepository.findById(id).orElse(null);
    }

    // Update Complaint Status
    public Complaint updateStatus(Long id, String status) {

        Complaint complaint = complaintRepository.findById(id).orElse(null);

        if (complaint == null) {
            return null;
        }

        complaint.setStatus(status);

        return complaintRepository.save(complaint);
    }

    // Delete Complaint
    public void deleteComplaint(Long id) {

        complaintRepository.deleteById(id);

    }

    // Get Complaints By User Id
    public List<Complaint> getComplaintsByUser(Long userId) {

        return complaintRepository.findByUserId(userId);

    }

    // Search Complaints By Title
    public List<Complaint> searchComplaints(String title) {

        return complaintRepository.findByTitleContainingIgnoreCase(title);

    }

    // Filter Complaints By Status
    public List<Complaint> filterByStatus(String status) {

        return complaintRepository.findByStatusIgnoreCase(status);

    }

}